From d627ff8250fdb7a092a75729b02c90cc75191640 Mon Sep 17 00:00:00 2001
From: Yihao Wu <wuyihao@linux.alibaba.com>
Date: Sun, 28 Mar 2021 01:59:28 +0800
Subject: [PATCH] Dynamic __schedule springboard

With this springboard, we don't have to customize the kernel's __schedule

Signed-off-by: Yihao Wu <wuyihao@linux.alibaba.com>
---
 kernel/sched/mod/core.c | 12 +++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

diff --git a/kernel/sched/mod/core.c b/kernel/sched/mod/core.c
index 5ec2ca5..3b8925a 100644
--- a/kernel/sched/mod/core.c
+++ b/kernel/sched/mod/core.c
@@ -2487,6 +2487,8 @@ asmlinkage __visible void schedule_tail(struct task_struct *prev)
 	calculate_sigpending();
 }
 
+extern unsigned long sched_springboard;
+
 /*
  * context_switch - switch to the new MM and the new thread's register state.
  */
@@ -2531,7 +2534,14 @@ context_switch(struct rq *rq, struct task_struct *prev,
 	prepare_lock_switch(rq, next, rf);
 
 	/* Here we just switch the register state and the stack. */
-	switch_to(prev, next, prev);
+	prepare_switch_to(next);
+	__asm__("pushq %0\n\t"
+		"jmp __switch_to_asm"
+		:
+		:"r"(sched_springboard), "D"(prev), "S"(next)
+	);
+
+	/* Below will not be executed, we'll return to vmlinux here */
 	barrier();
 
 	return finish_task_switch(prev);
2.20.1.2432.ga663e714

